src/server.rs

```diff
@@ -0,0 +1,512 @@
+    protocol: StreamProtocol,
+    sink: SinkState,
+    pending_outgoing_messages: Option<Vec<ProtoBlock>>,
+}
+
+impl<const S: usize> fmt::Debug for ServerConnectionHandler<S> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.write_str("ServerConnectionHandler")
+    }
+}
+
+#[derive(Default)]
+enum SinkState {
+    #[default]
+    None,
+    Requested,
+    Ready(Sink),
+}
+
+impl<const S: usize> ServerConnectionHandler<S> {
+    pub(crate) fn set_stream(&mut self, stream: libp2p_swarm::Stream) {
+        // Convert `AsyncWrite` stream to `Sink`
+        self.sink = SinkState::Ready(FramedWrite::new(stream, Codec));
+    }
+
+    pub(crate) fn queue_messages(&mut self, messages: Vec<(Vec<u8>, Vec<u8>)>) {
+        let block_list = messages
+            .into_iter()
+            .map(|(prefix, data)| ProtoBlock { prefix, data })
+            .collect::<Vec<_>>();
+
+        self.pending_outgoing_messages
+            .get_or_insert(Vec::with_capacity(block_list.len()))
+            .extend(block_list);
+    }
+
+    fn open_new_substream(
+        &mut self,
+    ) -> Poll<
+        ConnectionHandlerEvent<ReadyUpgrade<StreamProtocol>, StreamRequester, ToBehaviourEvent<S>>,
+    > {
+        self.sink = SinkState::Requested;
+
+        Poll::Ready(ConnectionHandlerEvent::OutboundSubstreamRequest {
+            protocol: SubstreamProtocol::new(
+                ReadyUpgrade::new(self.protocol.clone()),
+                StreamRequester::Server,
+            ),
+        })
+    }
+
+    fn poll_outgoing(
+        &mut self,
+        cx: &mut Context,
+    ) -> Poll<
+        ConnectionHandlerEvent<ReadyUpgrade<StreamProtocol>, StreamRequester, ToBehaviourEvent<S>>,
+    > {
+        loop {
+            match (&mut self.pending_outgoing_messages, &mut self.sink) {
+                (_, SinkState::Requested) => return Poll::Pending,
+                (None, SinkState::None) => return Poll::Pending,
+                (None, SinkState::Ready(sink)) => {
+                    if ready!(sink.poll_flush_unpin(cx)).is_err() {
+                        self.close_sink("poll_flush_unpin");
+                    }
+                    return Poll::Pending;
+                }
+                (Some(_), SinkState::None) => return self.open_new_substream(),
+                (pending_messages @ Some(_), SinkState::Ready(sink)) => {
+                    let messages = pending_messages
+                        .take()
+                        .expect("pending_messages can't be None here");
+                    let message = Message {
+                        payload: messages,
+                        ..Message::default()
+                    };
+
+                    if ready!(sink.poll_flush_unpin(cx)).is_err() {
+                        self.close_sink("poll_flush_unpin");
+                        continue;
+                    }
+
+                    if sink.start_send_unpin(&message).is_err() {
```

Comment:
I'm not sure if we need to do it in this PR but `Message` must be 4mb at max. So somehow we need to split messages to multiple ones.

Reply:
per daily, this will be handled separately for all sent messages in https://github.com/eigerco/beetswap/issues/24